packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -557,162 +558,118 @@ impl<'m> RunnerImpl<'m> {
+            },
+            DataType::UInt16 => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const u16, target_len);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            },
+            DataType::UInt32 => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const u32, target_len);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            },
+            DataType::Float64 => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const f64, target_len);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            },
+            DataType::Utf8 => unsafe {
+                // TODO: Use `data.len` or target_len?
+                //       (In practice, target_len has worked for a long time,
+                //       though that's not an ideal reason to use it. Maybe
+                //       `data.len` would also work.)
+                let offsets =
+                    slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&offsets));
+                let values = if let Some(size) = offsets.last() {
+                    slice::from_raw_parts(data.buffer_ptrs[1], *size as usize)
+                } else {
+                    &[]
+                };
+                builder = builder.add_buffer(values.into());
+            },
+            DataType::List(inner_field) => unsafe {
+                let offsets =
+                    slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&offsets));
+
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
+                    inner_field,
                     None,
-                )?])
-            }
-            DataType::FixedSizeList(field, multiplier) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+                )?);
+            },
+            DataType::FixedSizeList(inner_field, size) => {
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
-                    Some(data.len * multiplier as usize),
-                )?])
+                    inner_field,
+                    Some(data.len * *size as usize),
```

Comment:
See it also confuses me here as to why we're using `data.len` instead of `target_len`